pmd-lang-test/src/main/kotlin/net/sourceforge/pmd/test/AbstractMetricTestRule.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.test;
+
+import java.text.MessageFormat;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import net.sourceforge.pmd.RuleContext;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.metrics.Metric;
+import net.sourceforge.pmd.lang.metrics.MetricOption;
+import net.sourceforge.pmd.lang.metrics.MetricOptions;
+import net.sourceforge.pmd.lang.rule.AbstractRule;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+import net.sourceforge.pmd.properties.PropertyFactory;
+
+
+/**
+ * Abstract test rule for a metric. Tests of metrics use the standard
+ * framework for rule testing, using one dummy rule per metric. Default
+ * parameters can be overridden by overriding the protected methods of
+ * this class.
+ *
+ * @param <N> Result type of the metric. The nested subclasses provide
+ *            defaults for common result types
+ *
+ * @author Clément Fournier
+ */
+public abstract class AbstractMetricTestRule<N extends Number & Comparable<N>> extends AbstractRule {
```

Comment:
I think, that class can be probably moved to pmd-test module under src/main/java.